src/main/java/com/synopsys/integration/alert/channel/azure/boards/web/AzureOAuthCallbackController.java

```diff
@@ -72,73 +74,94 @@
+    private final OAuthRequestValidator oAuthRequestValidator;
+    private final AuthorizationManager authorizationManager;
 
     @Autowired
     public AzureOAuthCallbackController(ResponseFactory responseFactory, Gson gson, AzureBoardsChannelKey azureBoardsChannelKey,
         AzureBoardsCredentialDataStoreFactory azureBoardsCredentialDataStoreFactory, ProxyManager proxyManager, ConfigurationAccessor configurationAccessor,
-        AzureRedirectUtil azureRedirectUtil) {
+        AzureRedirectUtil azureRedirectUtil, OAuthRequestValidator oAuthRequestValidator, AuthorizationManager authorizationManager) {
         this.responseFactory = responseFactory;
         this.gson = gson;
         this.azureBoardsChannelKey = azureBoardsChannelKey;
         this.azureBoardsCredentialDataStoreFactory = azureBoardsCredentialDataStoreFactory;
         this.proxyManager = proxyManager;
         this.configurationAccessor = configurationAccessor;
         this.azureRedirectUtil = azureRedirectUtil;
+        this.oAuthRequestValidator = oAuthRequestValidator;
+        this.authorizationManager = authorizationManager;
     }
 
     @GetMapping
     public ResponseEntity<String> oauthCallback(HttpServletRequest request) {
         logger.debug("Azure OAuth callback method called");
+        if (!authorizationManager.hasExecutePermission(ConfigContextEnum.GLOBAL.name(), azureBoardsChannelKey.getUniversalKey())) {
+            logger.debug("Azure OAuth callback user does not have permission to call the controller.");
+            return responseFactory.createForbiddenResponse();
+        }
+
+        String state = request.getParameter("state");
         try {
             String requestURI = request.getRequestURI();
             String requestQueryString = request.getQueryString();
             logger.debug("Request URI {}?{}", requestURI, requestQueryString);
             String authorizationCode = request.getParameter("code");
-            String state = request.getParameter("state");
-            FieldAccessor fieldAccessor = createFieldAccessor();
-            if (fieldAccessor.getFields().isEmpty()) {
-                logger.error("Azure oauth callback: Channel global configuration missing");
+            if (!oAuthRequestValidator.hasRequestKey(state)) {
+                logger.info("OAuth request {} not found.", state);
             } else {
-                if (StringUtils.isBlank(authorizationCode)) {
-                    logger.error("Azure oauth callback: Authorization code isn't valid. Stop processing");
+                logger.info(createOAuthRequestLoggerMessage(state, "Processing..."));
+                oAuthRequestValidator.removeAuthorizationRequest(state);
+                FieldAccessor fieldAccessor = createFieldAccessor();
+                if (fieldAccessor.getFields().isEmpty()) {
+                    logger.error(createOAuthRequestLoggerMessage(state, "Azure oauth callback: Channel global configuration missing"));
```

Comment:
Could this be improved to an actual response? Or do we want to continue processing?

Reply:
Azure will be calling this controller since this is the callback URL.  So to send a response here would send it to the Azure server which would not update the Alert UI for the user.  We redirect back to the Alert UI at the end.  We could put query parameters with some details however we don't have a mechanism now to parse the query parameters in the URL to update the UI state at all.